alf/algorithms/merlin_algorithm.py

```diff
@@ -0,0 +1,441 @@
+# Copyright (c) 2019 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Implementation of MERLIN algorithm. See class MerlinAlgorithm for detail."""
+
+from collections import namedtuple
+import functools
+
+import tensorflow as tf
+import tensorflow_probability as tfp
+
+from tf_agents.agents.tf_agent import LossInfo
+from tf_agents.networks.network import Network
+from tf_agents.networks.actor_distribution_network import ActorDistributionNetwork
+from tf_agents.networks.value_network import ValueNetwork
+from tf_agents.specs.tensor_spec import TensorSpec
+from tf_agents.trajectories.policy_step import PolicyStep
+
+import alf
+from alf.algorithms.actor_critic_loss import ActorCriticLoss
+from alf.algorithms.memory import MemoryWithUsage
+from alf.algorithms.algorithm import Algorithm, AlgorithmStep
+from alf.algorithms.actor_critic_algorithm import ActorCriticInfo
+from alf.algorithms.on_policy_algorithm import OnPolicyAlgorithm
+from alf.algorithms.on_policy_algorithm import TrainingInfo
+from alf.algorithms.on_policy_algorithm import ActionTimeStep
+from alf.algorithms.vae import VariationalAutoEncoder
+from alf.utils.action_encoder import SimpleActionEncoder
+
+MBPState = namedtuple(
+    "MBPState",
+    [
+        "latent_vector",
+        "mem_readout",
+        "rnn_state",
+        "memory",  # memory state
+    ])
+
+MBPLossInfo = namedtuple("MBPLossInfo", ["decoder", "vae"])
+
+
+def get_rnn_cell_state_spec(cell):
+    """Get the state spec for RNN cell."""
+    return tf.nest.map_structure(
+        lambda size: TensorSpec(size, dtype=tf.float32), cell.state_size)
+
+
+def make_lstm_cell(lstm_size, name="lstm_cell"):
+    """Make a stacked LSTM cell.
+
+    Args:
+      lstm_size (int|list[int]): specifying the LSTM cell sizes to use.
+      name (str): name of the cell.
+      
+    Returns:
+      cell (LSTMCell|StackedRNNCell): the stacked LSTM cell.
+
+    """
+    implementation = 2  # FUSED IMPLEMENTATION
+    cells = [
+        tf.keras.layers.LSTMCell(
+            size, implementation=implementation, name=name + "/" + str(i))
+        for i, size in enumerate(lstm_size)
+    ]
+    if len(cells) == 1:
+        return cells[0]
+    else:
+        return tf.keras.layers.StackedRNNCells(cells)
+
+
+def _collect_variables(*modules):
+    modules = tf.nest.flatten(modules)
+    return sum([mod.variables for mod in modules], [])
+
+
+class MemoryBasedPredictor(Algorithm):
```

Comment:
everything should be annotated with @gin.configurable?

Reply:
good idea